tests/test_handlers.py

```diff
@@ -562,6 +563,83 @@ def test_am1_differences():
             assert 0
 
 
+def test_compute_or_load_am1_charges():
+    """Loop over test ligands, asserting that charges are stored in expected property and that the same charges are
+    returned on repeated calls"""
+
+    # get some molecules
+    cache_key = "AM1Cache"
+    suppl = Chem.SDMolSupplier("tests/data/ligands_40.sdf", removeHs=False)
+    all_mols = [mol for mol in suppl]
+    all_mols = all_mols[:5]  # truncate so that whole test is ~ 10 seconds
+
+    # don't expect AM1 cache yet
+    for mol in all_mols:
+        assert not mol.HasProp(cache_key)
+
+    # compute charges once
+    fresh_am1_charges = [nonbonded.compute_or_load_am1_charges(mol) for mol in all_mols]
+
+    # expect each mol to have AM1 cache now
+    for mol in all_mols:
+        assert mol.HasProp(cache_key)
+
+    # expect the same charges as the first time around
+    cached_am1_charges = [nonbonded.compute_or_load_am1_charges(mol) for mol in all_mols]
+    for (fresh, cached) in zip(fresh_am1_charges, cached_am1_charges):
+        np.testing.assert_allclose(fresh, cached)
+
+
+def test_bond_smirks_matches():
+    """Loop over test ligands, asserting that
+    * returned indices are in bounds
+    * returned bonds are present in the mol"""
+    # get some molecules
+    suppl = Chem.SDMolSupplier("tests/data/ligands_40.sdf", removeHs=False)
+    all_mols = [mol for mol in suppl]
+
+    # get some bond smirks
+    smirks_list = [smirks for (smirks, param) in AM1CCC_CHARGES["patterns"]]
+
+    for mol in all_mols:
+        bond_idxs, type_idxs = nonbonded.bond_smirks_matches(mol, smirks_list)
+
+        # assert indices in bounds
+        assert (bond_idxs.min() >= 0) and (bond_idxs.max() < mol.GetNumAtoms())
+        assert (type_idxs.min() >= 0) and (type_idxs.max() < len(smirks_list))
+
+        # assert that bond_idxs are present in the mol
+        bonds = set()
+        for bond in mol.GetBonds():
+            a, b = bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()
+            bonds.add((a, b))
+            bonds.add((b, a))
+        for bond in bond_idxs:
+            assert tuple(bond) in bonds
+
+
+def test_apply_bond_charge_corrections():
+    """Assert that applying random bond charge corrections does not change net charge"""
+
+    n_random_tests = 100
+    np.random.seed(2022)
+
+    for _ in range(n_random_tests):
+        n_atoms = np.random.randint(10, 50)
+        initial_charges = np.random.randn(n_atoms)
+        initial_net_charge = np.sum(initial_charges)
+
+        # arbitrary: duplicates okay, reversals okay, symmetry not required
+        n_directed_bonds = np.random.randint(50, 100)
+        bonds = np.random.randint(0, n_atoms, size=(n_directed_bonds, 2))
+        deltas = np.random.randn(n_directed_bonds)
+
+        final_charges = nonbonded.apply_bond_charge_corrections(initial_charges, bonds, deltas)
+        final_net_charge = np.sum(final_charges)
+
+        np.testing.assert_almost_equal(final_net_charge, initial_net_charge)
```

Comment:
Would it be worth also checking that some charges are changed?

Reply:
Agreed -- addressed in https://github.com/proteneer/timemachine/pull/571/commits/ae9ebec7f37beb632c4aee98b42be2b3753c9a52 !